File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;

        int dp[] = new int[n];
        Arrays.fill(dp, 1);

        for (int i=0; i < n; i++) {
            for (int j=0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    dp[i] = Math.max(dp[j] + 1, dp[i]);
                }
            }
        }

        int max = dp[0];
        System.out.print(dp[0] + " ");
        for (int i=1; i < n; i++) {
            max = Math.max(max, dp[i]);
            System.out.print(dp[i] + " ");
        }

        return max;

        // int dp[][] = new int[n + 1][n + 1];

        // for (int i=1; i <= n; i++) {
        //     for (int j=1; j <= i; j++) {

        //     }
        // }

        // for (int i=1; i <= n; i++) {
        //     for (int j=1; j <= n; j++) {
        //         System.out.print(dp[i][j] + " ");
        //     }

        //     System.out.println():
        // }

    }
    static int LIS(int nums[], int idx, int last) {
        // base case
        if (idx == nums.length) return 0;
        
        // write a recurrunce

        int take = 0, notTake = 0;
        if (last == -1 || nums[last] < nums[idx]) {
            take = 1 + LIS(nums, idx + 1, idx);
        }
        notTake = LIS(nums, idx + 1, last);

        return Math.max(take, notTake);
    }
}

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        ans.push_back(nums[0]);
        int len = 1;

        for (int i=0; i < n; i++) {
            if (nums[i] > ans.back()) {
                ans.push_back(nums[i]);
                len ++;
            }
             else {
                int idx = lower_bound(ans.begin(), ans.end(), nums[i]) - ans.begin();
                ans[idx] = nums[i];
            }

        }

        return ans.size();
        // return len;
    }
};